[autoconf] inotify broken on FreeBSD; --disable-inotify not working

## Steps to Reproduce

1. Any FreeBSD 11.3+ will do ya...
2. Use https://reviews.freebsd.org/D23300 on your ports tree
3. If `libinotify` is installed on the system, `autoconf` will ignore `--disable-inotify` 
4. If `libinotify` is installed (or not installed) on the system, and is desired, `autoconf` will fail to pull in correctly as dotnet/corefx#40202 was not fully ported (missing the cmake fixes.)

### Current Behavior

a. `autoconf` ignores --disable-inotify and aggressively detects on FreeBSD
b. `autoconf` does not compile correctly with `libinotify`

```
./configure --with-csc=mcs --disable-btls --disable-dtrace --disable-boehm --disable-inotify --disable-ninja --disable-nls --with-x --prefix=/usr/local --localstatedir=/var --mandir=/usr/local/man --disable-silent-rules --infodir=/usr/local/share/info/ --build=aarch64-portbld-freebsd13.0
...
checking sys/inotify.h usability... no
checking sys/inotify.h presence... no
checking for sys/inotify.h... no
checking for inotify_init in -linotify... yes
checking for inotify_rm_watch with unsigned wd... no
checking for inotify_init... yes
checking for inotify_add_watch... yes
checking for inotify_rm_watch... yes


gmake[5]: Entering directory '/usr/ports/lang/mono/work/mono-6.8.0.96/mono/native'
../../doltlibtool  --tag=CC   --mode=compile cc -DHAVE_CONFIG_H -I. -I../..  -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Qunused-arguments -Wno-unused-function -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -DLIBICONV_PLUG -DGC_FREEBSD_THREADS -DHOST_BSD -D_WITH_GETLINE -g  -DUSE_COMPILER_TLS -I../.. -I../../mono  -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Qunused-arguments -Wno-unused-function -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -O2 -pipe  -DLIBICONV_PLUG -fstack-protector-strong -fno-strict-aliasing  -Wmissing-prototypes -Wstrict-prototypes -Wnested-externs -Wno-format-zero-length -std=gnu99 -fno-strict-aliasing -fwrapv -DMONO_DLL_EXPORT -g -U_FORTIFY_SOURCE -I/usr/ports/lang/mono/work/mono-6.8.0.96/external/corefx/src/Native/Unix/Common -I/usr/ports/lang/mono/work/mono-6.8.0.96/external/corefx/src/Native/Unix/System.Native -I/usr/ports/lang/mono/work/mono-6.8.0.96/external/corefx/src/Native/AnyOS/brotli/include -I../../mono/eglib -I../../mono/eglib -fvisibility=hidden -DBROTLI_BUILD_NO_RBIT -Wno-typedef-redefinition  -MT libmono_native_la-pal_io.lo -MD -MP -MF .deps/libmono_native_la-pal_io.Tpo -c -o libmono_native_la-pal_io.lo `test -f '../../external/corefx/src/Native/Unix/System.Native/pal_io.c' || echo './'`../../external/corefx/src/Native/Unix/System.Native/pal_io.c
../../external/corefx/src/Native/Unix/System.Native/pal_io.c:37:10: fatal error: 'sys/inotify.h' file not found
#include <sys/inotify.h>
         ^~~~~~~~~~~~~~~
1 error generated.
gmake[5]: *** [Makefile:2119: libmono_native_la-pal_io.lo] Error 1
gmake[5]: Leaving directory '/usr/ports/lang/mono/work/mono-6.8.0.96/mono/native'
```

### Expected Behavior

Should respect `--disable-inotify` (rely on kqueue) and should also correctly detect the location of `inotify.h`

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[ ] Windows
[X] FreeBSD (ports)

**Version Used**:

6.8.0.96 (dl.mono-project.org tarball)

## Stacktrace

N/A

Also missed the upper-level error (earlier in process): 

```
filewatcher.c:61:10: fatal error: 'sys/inotify.h' file not found
#include <sys/inotify.h>
         ^~~~~~~~~~~~~~~
```

Many thanks to @akoeplinger for his help when autoconf kicked my butt. (Pointy-hat to: rootwyrm, who used notes from a wildly wrong version.) We now have a patch suitable for FreeBSD ports to give us desired `--disable-inotify` behavior.

This also helped me identify more clearly what the specific problem is; it actually appears to be an autotools vs. cmake issue. cmake behaves better on FreeBSD and correctly identifies the additional requirement for `-I/usr/local/include` when used in CoreFX. Which actually helped me to identify _another_ problem in autoconf.

`./configure --includedir=/usr/include --includedir=/usr/local/include --disable-dtrace --disable-boehm --disable-ninja --with-x --prefix=/usr/local --localstatedir=/var --mandir=/usr/local/man --disable-silent-rules --infodir=/usr/local/share/info/ --build=amd64-portbld-freebsd12.1` 

This should produce a working result, we would think. Instead, autotools does this:
```
configure:27323: clang -o conftest -Wall -Wunused -Wmissing-declarations -Wpointer-arith -Wno-cast-q
ual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Qunused-argument
s -Wno-unused-function -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assign -Wno-ret
urn-stack-address -Wno-constant-logical-operand -Wno-zero-length-array -g -O2 -Wmissing-prototypes -
Wstrict-prototypes -Wnested-externs -Wno-format-zero-length -std=gnu99 -fno-strict-aliasing -fwrapv
-DMONO_DLL_EXPORT -g -Werror=sign-conversion -Wall -Wunused -Wmissing-declarations -Wpointer-arith -
Wno-cast-qual -Wwrite-strings -Wno-switch -Wno-switch-enum -Wno-unused-value -Wno-attributes -Qunuse
d-arguments -Wno-unused-function -Wno-tautological-compare -Wno-parentheses-equality -Wno-self-assig
n -Wno-return-stack-address -Wno-constant-logical-operand -Wno-zero-length-array  -DGC_FREEBSD_THREA
DS -DHOST_BSD -D_WITH_GETLINE -g   -lutil -pthread -L/usr/local/lib conftest.c -lm -lintl   -linotif
y -pthread >&5
conftest.c:313:12: fatal error: 'sys/inotify.h' file not found
                #include <sys/inotify.h>
                         ^~~~~~~~~~~~~~~
```

Notice the lack of `-I/usr/local/include`; that's the problem. So it's looking only in `/usr/include` which is why it can't find `sys/inotify.h` (which as you guessed, lives at `/usr/local/include/sys/inotify.h`). Which cmake automatically handles, but autotools seems not to handle correctly. This reproduced when run via `./autogen.sh` as well. 
I took an initial swing at bodging the include in there, but it didn't work as expected, so I'm not sure what I'm missing here. (Other than a lot of warnings about autotools not really handling this sort of need gracefully.) And since this is a more significant change to the autotools portion, I'd definitely like to get some feedback from more experienced folks on the correct way to address the root cause.

--- a/configure.ac
+++ b/configure.ac
@@ -2372,14 +2372,6 @@ if test x$host_win32 = xno; then
 	dnl *****************************
 	AC_CHECK_LIB(socket, socket, LIBS="$LIBS -lsocket")
 
-	case "$host" in
-		*-*-*freebsd*)
-			dnl *****************************
-			dnl *** Checks for libinotify ***
-			dnl *****************************
-			AC_CHECK_LIB(inotify, inotify_init, LIBS="$LIBS -linotify")
-	esac
-
 	dnl *****************************
 	dnl *** Checks for IPPROTO_IP ***
 	dnl *****************************
@@ -3647,6 +3639,34 @@ if test x$host_win32 = xno; then
 	], [
 		AC_MSG_RESULT(no)
 	])
+
+	case "$host" in
+		*-*-*freebsd*)
+			dnl *****************************
+			dnl *** Checks for libinotify ***
+			dnl *****************************
+			AC_CHECK_LIB(inotify, inotify_init, LIBS="$LIBS -linotify")
+			if test "x$ac_cv_lib_inotify_inotify_init" = "xyes" ; then
+				AC_DEFINE(HAVE_LIBINOTIFY, 1, [FreeBSD libinotify kqueue shim])
+				dnl Needs to be done this way to avoid collision with various 
+				dnl ports including glib and llvm*
+				METADATA_CFLAGS="-I/usr/local/include"
+			fi
+			dnl Workaround due to inotify_rm_watch check failing without -I
+			AC_MSG_CHECKING(for inotify_rm_watch with unsigned wd in libinotify)
+			AC_TRY_LINK([
+				#include </usr/local/include/sys/inotify.h>
+			], [
+				intptr_t fd;
+				uint32_t wd;
+				int result = inotify_rm_watch(fd, wd);
+			],[
+			   AC_MSG_RESULT(yes)
+			   AC_DEFINE(INOTIFY_RM_WATCH_WD_UNSIGNED, 1, [inotify_rm_watch with unsigned wd])
+			], [
+				AC_MSG_RESULT(no)
+			])
+	esac
 
 	CFLAGS="$ORIG_CFLAGS"
 
@@ -6565,7 +6585,7 @@ elif case $host_os in freebsd*) true;; *) false;; esac; then
 	MONO_NATIVE_CXX=$CXX
 	MONO_NATIVE_CPPFLAGS=$CPPFLAGS
 	MONO_NATIVE_CXXFLAGS=$CXXFLAGS
-	MONO_NATIVE_CFLAGS=$CFLAGS
+	MONO_NATIVE_CFLAGS="$CFLAGS -I/usr/local/include"
 	MONO_NATIVE_LDFLAGS=$LDFLAGS
 
 	mono_native=yes

--- a/configure.ac
+++ b/configure.ac
@@ -3651,6 +3651,7 @@ if test x$host_win32 = xno; then
 				dnl Needs to be done this way to avoid collision with various 
 				dnl ports including glib and llvm*
 				METADATA_CFLAGS="-I/usr/local/include"
+				AC_SUBST(METADATA_CFLAGS)
 			fi
 			dnl Workaround due to inotify_rm_watch check failing without -I
 			AC_MSG_CHECKING(for inotify_rm_watch with unsigned wd in libinotify)

--- a/mono/metadata/Makefile.am
+++ b/mono/metadata/Makefile.am
@@ -107,7 +107,7 @@ noinst_LTLIBRARIES = libmonoruntime-config.la $(support_libraries) $(boehm_libra
 
 lib_LTLIBRARIES = $(icall_table_libraries) $(ilgen_libraries)
 
-AM_CPPFLAGS = -I$(top_srcdir) -I$(top_srcdir)/mono $(LIBGC_CPPFLAGS) $(GLIB_CFLAGS) $(SHARED_CFLAGS)
+AM_CPPFLAGS = -I$(top_srcdir) -I$(top_srcdir)/mono $(LIBGC_CPPFLAGS) $(GLIB_CFLAGS) $(SHARED_CFLAGS) $(METADATA_CFLAGS)
 
 #
 # Make sure any prefix changes are updated in the binaries too.

